### PR TITLE
[PQ] Hide FNX clusters in DiscoverClusters unless :authority matches a Balancer (LOGBROKER-10621)

### PR DESCRIPTION
**Pull request by nshestakov**
[LOGBROKER-10621](https://st.yandex-team.ru/LOGBROKER-10621)

### Changelog entry 

Topics: Cluster Discovery returns FNX clusters only to clients whose gRPC `:authority` matches a `Balancer` row.

### Description for reviewers 

Ordinary clients must not see FNX endpoints. ClusterTracker keeps the full Cluster list; CDS and PQv0 filter per request.

**Schema**
- Auto-migrate `Cluster.fnx` and create `Balancer(name, clusters Utf8 CSV of FNX names)`.
- Combined tracker version is Cluster version + Balancer version.
- Remigrate only when the table or column is missing; ignore already-exists DDL errors; fail closed if Balancer cannot be listed.

**Visibility**
- Non-FNX clusters are always returned.
- Extra FNX names come from the `Balancer` row whose `name` matches normalized gRPC `:authority` (lowercase, strip port, IPv6-aware).
- Empty or unknown host → only non-FNX.

**PQv0 / SDK**
- PQv0 drops its duplicate KQP Cluster poller and subscribes to ClusterTracker, then applies the same host filter.
- C++ SDK sends CDS RPCs with `UseDiscoveryEndpoint` so `:authority` stays the client VIP when `Database` is set. No hardcoded FNX balancer name in `IsFederation`.

**Tests**
- Baseline UT pins pre-FNX DiscoverClusters/ClusterTracker behavior.
- Feature UT: migrate old schema, accept pre-migrated schema, hide FNX unless balancer matches.
- Extra UT for tracker/CDS edge cases (waiting, wakeup, remigrate after DROP, empty list, health).
- Compatibility: DiscoverClusters stays SUCCESS on rolling/restart of new ydbd onto the old Cluster schema.

**Out of this PR**
- FDS / `kikimr/yndx` Config Manager projection of `Cluster.fnx` and `Balancer`.

### CODE REVIEW

File: ydb/core/persqueue/public/cluster_tracker/cluster_tracker.cpp [hunk 1]
```diff
@@ -202,51 +163,205 @@ class TClusterTracker: public TActorBootstrapped<TClusterTracker> {
...
+                        NYdb::TResultSetParser parser(record.GetResponse().GetYdbResults(0));
+                        UpdateBalancersList(parser);
+                    } else if (PendingClustersList) {
+                        FinishClustersList(std::move(PendingClustersList));
+                    }
+                    return;
+                }
+                {
+                    const bool remigrate = IssuesLookLikeClusterSchemaGone(issues);
+                    YDB_LOG_ERROR_CTX(Ctx(), "Failed to list balancers, publish clusters with empty balancer cache",
```
[the remark this answers is not in the document]
Comment (nshestakov) [reply ydb/core/persqueue/public/cluster_tracker/cluster_tracker.cpp]: Fixed in 55d1fbabcc2: on `ListBalancers` failure we republish with the last successful `BalancerVersion` and balancer cache. `DiscoverClusters.version` is reset only after a successful `ListBalancers`.

[the remark this answers is not in the document]
Comment (nshestakov) [reply ydb/core/persqueue/public/cluster_tracker/cluster_tracker.cpp]: Duplicate of the `ListBalancers` version-regression note; fixed in 55d1fbabcc2 as in the sibling thread.

File: ydb/core/persqueue/public/cluster_tracker/cluster_select.cpp [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+}
+
+TString MakeListBalancersQuery(TStringBuf balancerTablePath, TStringBuf versionTablePath) {
+    return Sprintf(
+        R"(
+               --!syntax_v1
+               SELECT B.name, B.clusters, V.version FROM `%s` AS B
+               CROSS JOIN
+               (SELECT MAX(version) AS version FROM `%s` WHERE name == 'Balancer') AS V;
+            )", TString(balancerTablePath).c_str(), TString(versionTablePath).c_str());
```
[the remark this answers is not in the document]
Comment (nshestakov) [reply ydb/core/persqueue/public/cluster_tracker/cluster_select.cpp:108]: Fixed in 55d1fbabcc2: `CREATE TABLE Cluster` now includes `kikimrHost Utf8` to match federation_recipe. Together with list-first, we also skip creating Cluster when the table already exists.

File: ydb/core/persqueue/public/cluster_tracker/cluster_tracker.cpp [hunk 3]
```diff
@@ -202,51 +163,205 @@ class TClusterTracker: public TActorBootstrapped<TClusterTracker> {
...
         req->Record.MutableRequest()->SetUsePublicResponseDataFormat(true);
-        // useless without explicit session
-        // req->Record.MutableRequest()->MutableQueryCachePolicy()->set_keep_in_cache(true);
         req->Record.MutableRequest()->MutableTxControl()->mutable_begin_tx()->mutable_serializable_read_write();
         req->Record.MutableRequest()->MutableTxControl()->set_commit_tx(true);
-
         Send(NKqp::MakeKqpProxyID(Ctx().SelfID.NodeId()), req.Release());
     }
 
+    void StartMigrateOrList() {
```
[the remark this answers is not in the document]
Comment (nshestakov) [reply ydb/core/persqueue/public/cluster_tracker/cluster_tracker.cpp:199]: Fixed in 55d1fbabcc2: the tracker lists first. DDL runs only when `ListClusters` / `ListBalancers` see a missing table (Cluster chain, Balancer only, or Versions only). A healthy schema no longer gets CREATE/ALTER on startup.

File: ydb/services/deprecated/persqueue_v0/grpc_pq_clusters_updater_actor.cpp [hunk 4]
```diff
@@ -1,42 +1,20 @@
...
 
-static const int CLUSTERS_UPDATER_TIMEOUT_ON_ERROR = 1;
-
-
 TClustersUpdater::TClustersUpdater(IPQClustersUpdaterCallback* callback, TStatus::TPtr& status)
     : Callback(callback)
     , Status(status)
     {};
 
 void TClustersUpdater::Bootstrap(const NActors::TActorContext& ctx) {
```
[the remark this answers is not in the document]
Comment (nshestakov) [reply ydb/services/deprecated/persqueue_v0/grpc_pq_clusters_updater_actor.cpp:11]: Won't change. Federation / CDS deployments set `PQConfig.ClusterTablePath`. PQv0 with only `Root` and no cluster-discovery config is not a supported combination for this change.

File: ydb/services/deprecated/persqueue_v0/grpc_pq_read.cpp [hunk 5]
```diff
@@ -26,9 +26,9 @@ void TPQReadService::TSession::OnCreated() {
         return;
     }
     // Create actor for current session.
-    auto clusters = Proxy->GetClusters();
+    auto clusters = Proxy->GetClusters(GetAuthority());
     auto localCluster = Proxy->GetLocalCluster();
-    if (NeedDiscoverClusters && (clusters.empty() || localCluster.empty())) {
+    if (NeedDiscoverClusters && (localCluster.empty() || !Proxy->HasClustersList())) {
```
[the remark this answers is not in the document]
Comment (nshestakov) [reply ydb/services/deprecated/persqueue_v0/grpc_pq_read.cpp:31]: Intended. A loaded list that is empty for this `:authority` (all clusters FNX, unknown host) should not look like INITIALIZING — the tracker is healthy, the client is simply not allowed to see those DCs. Serving the empty visible set matches DiscoverClusters.

File: ydb/core/persqueue/public/cluster_tracker/cluster_tracker.cpp [hunk 6]
```diff
@@ -202,51 +163,205 @@ class TClusterTracker: public TActorBootstrapped<TClusterTracker> {
...
         const auto& record = ev->Get()->Record;
-        if (record.GetYdbStatus() == Ydb::StatusIds::SUCCESS) {
-            NYdb::TResultSetParser parser(record.GetResponse().GetYdbResults(0));
-            if (parser.RowsCount()) {
-                YDB_LOG_DEBUG_CTX(Ctx(), "HandleWhileWorking TEvQueryResponse UpdateClustersList");
-                UpdateClustersList(parser);
+        const TString issues = record.ShortDebugString();
+        const bool success = record.GetYdbStatus() == Ydb::StatusIds::SUCCESS;
+        const EQueryKind kind = CurrentQuery;
+        QueryInFlight = false;
```
[the remark this answers is not in the document]
Comment (nshestakov) [reply ydb/core/persqueue/public/cluster_tracker/cluster_tracker.cpp:232]: Skipping. Unreachable with a well-behaved KQP proxy; not worth extra correlation state on the request.

File: ydb/core/persqueue/public/cluster_tracker/cluster_select.cpp [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+    }
+
+    if (host.StartsWith('[')) {
+        const auto end = host.find(']');
+        if (end != TString::npos) {
+            return host.substr(1, end - 1);
+        }
+    }
+
+    const auto firstColon = host.find(':');
```
[the remark this answers is not in the document]
Comment (nshestakov) [reply ydb/core/persqueue/public/cluster_tracker/cluster_select.cpp:55]: Won't change. CM writes `Balancer.name` as the VIP hostname without a trailing dot; that is the contract for `:authority` after port/case normalization.

File: ydb/core/persqueue/public/cluster_tracker/cluster_select.cpp [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+}
+
+bool IssuesLookLikeAlreadyExists(TStringBuf issues) {
+    return issues.Contains("already exists")
+        || issues.Contains("Already exists")
+        || issues.Contains("duplicate")
+        || issues.Contains("Duplicate");
+}
+
+bool IssuesLookLikeMissingTable(TStringBuf issues) {
```
[the remark this answers is not in the document]
Comment (nshestakov) [reply ydb/core/persqueue/public/cluster_tracker/cluster_select.cpp:183]: Won't change. After list-first, this matcher only starts CREATE when list already failed, and we no longer run `UPDATE fnx`. A false positive remigrate is `CREATE IF NOT EXISTS` plus another list — noisy, not a correctness issue.

File: ydb/core/persqueue/public/cluster_tracker/cluster_list.cpp [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    ClustersByBalancer.clear();
+
+    DefaultVisibleClusters.reserve(Clusters.size());
+    for (const auto& cluster : Clusters) {
+        if (!cluster.IsFnx) {
+            DefaultVisibleClusters.push_back(cluster);
+        }
+    }
+
+    for (const auto& [host, extraFnxNames] : Balancers) {
```
[the remark this answers is not in the document]
Comment (nshestakov) [reply ydb/core/persqueue/public/cluster_tracker/cluster_list.cpp:28]: Won't change. `Balancer.name` is written by CM as the VIP hostname; we match `:authority` after port/case normalization against that contract.

File: ydb/tests/compatibility/topics/test_cluster_discovery.py [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+    for path in ("/Root/PQ", "/Root/PQ/Config", "/Root/PQ/Config/V2"):
+        try:
+            driver.scheme_client.make_directory(path)
+        except Exception as exc:
+            logger.info("mkdir %s: %r", path, exc)
+
+    with ydb.QuerySessionPool(driver) as session_pool:
+        session_pool.execute_with_retries(
+            f"""
+            CREATE TABLE `{CLUSTER_TABLE}` (
```
[the remark this answers is not in the document]
Comment (nshestakov) [reply ydb/tests/compatibility/topics/test_cluster_discovery.py]: Fixed in 4f0a4c72eb6: seed now uses `CREATE TABLE IF NOT EXISTS` for Cluster and Versions, then UPSERT. Either the tracker or the test can create first.

**Review by alexnick88 — APPROVED** [review]

**Review by Shfdis — APPROVED** [review]
Sdk part lgtm
